src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,176 @@
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> first;
+    private Node<T> last;
+    private int size;
+
     @Override
     public boolean add(T value) {
-        return false;
+        Node<T> currentNode = last;
+        Node<T> newNode = new Node<>(currentNode, value, null);
+        last = newNode;
+        if (currentNode == null) {
+            first = newNode;
+        } else {
+            currentNode.next = newNode;
+        }
+        size++;
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
+        Node<T> currentNode = findElement(index);
+        checkIndex(index, index > size);
+        if (index == 0) {
+            addFirst(value, currentNode);
+            return;
+        }
+        if (index == size) {
+            add(value);
 
+        } else {
+            Node<T> pred = currentNode.prev;
+            Node<T> newNode = new Node<>(pred, value, currentNode);
+            currentNode.prev = newNode;
+            if (pred == null) {
+                first = newNode;
+            } else {
+                pred.next = newNode;
+            }
+            size++;
+        }
+    }
+
+    private void addFirst(T value, Node<T> currentNode) {
+        Node<T> newNode = new Node<>(null, value, currentNode);
+        first = newNode;
+        if (currentNode == null) {
+            last = newNode;
+        } else {
+            currentNode.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        if (list == null) {
+            return false;
+        }
+        for (T t : list) {
+            add(t);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index, index >= size);
+        return findElement(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index, index >= size);
+        Node<T> newNode = findElement(index);
+        T oldValue = newNode.item;
+        newNode.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index, index >= size);
+        Node<T> currentNode = findElement(index);
+        Node<T> oldValue = currentNode;
+        if (index == 0) {
+            removeFirst(currentNode);
+            return oldValue.item;
+        }
+        if (index == size - 1) {
+            removeLast(currentNode);
+            return oldValue.item;
+        }
+        currentNode.prev.next = oldValue.next;
+        currentNode.next.prev = oldValue.prev;
+        size--;
+        return oldValue.item;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = first;
+        for (int i = 0; i < size; i++) {
+            if (currentNode.item == null || currentNode.item.equals(object)) {
```

Comment:
what does happen if currentNode.item == null and object != null?

Reply:
done